### ISSUE TITLE
Придумать новый скролл

### ISSUE THREAD

**Issue by krauklish** [RusPosevkin/evolution#34]
Что понятно: наши варианты горизонтального скролла не катят. Очевидно, он конфликтует с нативным горизонтальным скроллом и не годится для трекпадов, тачпадов, мэджикмаусов и всякого такого. 

В идеальном варианте нужен скролл, который по колесу мыши будет скролить горизонтально, а по трекпаду — по умолчанию. Насколько я понимаю, средствами современного js это просто невозможно сделать. Но может я не прав, и какой-то способ есть? @RusPosevkin, надежда на тебя.

А я пока подумаю, как еще можно было бы скролить, чтоб всем было удобно.

**Comment by RusPosevkin** [comment]
В JS мы не можем отличить трекпад используется или мышь

**Comment by krauklish** [comment]
Да, это я уже осознал. 

Я вот думал про наши скрипты, которые меняли направление скрола. Почему они подглючивают? Мне кажется, потому что инвертированный горизонтальный скролл конфликтует с нативным.

Но может быть можно инвертировать оба направления? То есть вертикальный скрол по трекпаду дает горизонтальный скрол страницы. И наоборот: горизонтальный скрол трекпада дает вертикальный скрол страницы. Тогда они ведь не будут конфликтовать, так?

Короче, во-первых, я не понимаю, прав ли я в целом, а во-вторых, не знаю, можно ли это реализовать.

**Comment by RusPosevkin** [comment]
Как уже говорил, вмешательство в работу нативного скролла – плохая практика. Если сделаем #32, то и выпилим скрипт с инверсией скролла. У не-mac пользователя  и так будет удобная возможность пользоваться сайтом на десктопе

**Comment by RusPosevkin** [comment]
В https://github.com/RusPosevkin/evolution/commit/d06557fe1706b714139ee0bdbd2a66d229891b18 у нас появился классный скролл. 
Глючный скрипт больше не нужен – в https://github.com/RusPosevkin/evolution/commit/937e4d571171a044e427f588ec965b6691f5c985 я его выпилил.

**Comment by krauklish** [comment]
Блин, я на самом деле очень страдаю от того, что у нас нельзя скроллить колесом мышки :( 

Тесты показали, что ребята на винде вообще не врубаются, как скролить. Классный скролл правда классный, но он все-таки слегка сложный. 

Я согласен, что ломать нативный скрол — обычно плохая идея. Но когда у нас однозначно горизонтальное направление страницы, лучше бы его инвертировать. Точно не получится то, что я предложил: инвертировать оба направления, чтобы не было конфликта на трекпаде?

**Comment by RusPosevkin** [comment]
Мы попробовали 2 библиотеки с инвертированием, они обе оказались глючные в случае маковского скролла. 

Сама задача такого инвертирования нетривиальная. А учитывая сколько интерфейсных решений уже сейчас реализовано с учетом скроллинга (и сколько костылей ради этого запилено), то сложность растет экспоненциально.

Тут даже если прямо сейчас найти готовое решение, которое не будет глючить, как предыдущие, то уже потребуется приличное количество времени, чтобы интегрировать ее в текущее состояние.

Большинство мышей сейчас вполне поддерживают горизонтальный скролл. Если нет – есть интерфейсный скролл. И, возможно, стоит подумать, как более наглядно показать пользователю, что вот этот элемент – скролл. На мой взгляд, сейчас это не совсем очевидно.

**Comment by krauklish** [comment]
Там нужно какое-то аккуратное дизайн-решение, но у меня просто не хватает на него времени сейчас. 


Проблема-то в том, что кликать и таскать — это в сто раз неудобнее, чем скролить. Я реально подумываю, что наше глючное решение было не таким и плохим: ну да, подергивается на трекпаде, но блин, на одного человека с трекпадом приходится десять с обычной мышкой. И если на трекпаде подключивает, то на мышке вообще ничего не работает.


>Большинство мышей сейчас вполне поддерживают горизонтальный скролл.

А как они работают?

**Comment by RusPosevkin** [comment]
На таких мышах обычно стрелочки горизонтальные рисуют. Просто сдвигаешь колесико по горизонтали в нужную сторону и отрабатывает горизонтальный скролл.
![a982bf](https://user-images.githubusercontent.com/1483747/27052256-cc1d12c8-4fc1-11e7-8ef1-8dd1a34fab3a.jpg)

На самом деле, это типичная ситуация для пользователя с мышкой при работе с сайтом с горизонтальной прокруткой. Ты просто берешь и скроллишь с помощью полосы прокрутки. Здесь же мы вырубили нативную полосу прокрутки и запилили свою. За счет большей высоты, по сравнению с нативом, в нее удобнее попасть и в целом довольно удобно работать.

А решение с глючной библиотекой, что было, это не просто немного подергивается. Это реально очень бесит. Я бы сразу ушел с такого сайта(

**Comment by krauklish** [comment]
О, у меня на мышке есть такое, но оно не скроллит  :)

>А решение с глючной библиотекой, что было, это не просто немного подергивается. Это реально очень бесит. Я бы сразу ушел с такого сайта(

Ну вот я думаю, что надо допилить стороннюю библиотеку так, чтобы еще и горизонтальный скролл инвертировался в вертикальный. Реально ведь в этом дело, нет? На трекпаде конфликт двух скроллов.

Я понимаю, что это затратно. Это так, праздные размышления.

**Comment by RusPosevkin** [comment]
Прямо сейчас нет под рукой мышки, чтобы проверить. Но еще недавно скроллилось.

Когда вторая библиотека начала глючить, я посидел поразбирался с тем, как ее пофиксить. Понял, что это гиблое дело. В лучшем случае на доработку этой библиотеки потребуется столько времени, сколько затрачено на разработку всего этого проекта. Я не считаю это рациональным расходованием времени.

**Comment by krauklish** [comment]
Ох, блин :) Ладно, я сдаюсь, бросаем это инвертирование.

Посмотрим, что покажет метрика после релиза. Если будет видно, что у людей проблемы, подумаю над этим еще.

**Comment by krauklish** [comment]
У меня тут очередная идея: не добавить ли нам таскание ко всей странице? Чтоб можно было мышкой скроллить, как на телефоне. Как думаешь?

**Comment by krauklish** [comment]
Я уже несколько откликов получил, что скролл не работает :)

Надо, короче, опробовать библиотечку, которую обсуждали. По-моему, вот это она: https://github.com/jquery/jquery-mousewheel

Переоткрою, чтобы не забыть.

**Comment by krauklish** [comment]
Так, я почитал-потыкал и понял, что нет, библиотека-то не для этого, а просто для нормализации дельты по браузерам. А для собственно скролла там буквально десять строчек кода. 

Я сейчас добавил эту штуку, просто чтобы посмотреть, как она будет работать: f68c007b2a1bf9edccc76a78967d96efbce7d4c9

Уже вижу, что на винде скорость примерно раз в десять меньше, чем на маке, то есть решение, мягко говоря, неокончательное. 

Но с этим реально надо что-то придумать, потому что прям проблемы у людей.
